Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,447 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+
+import Darwin
+import CNIOLinux
+import CoreFoundation
+
+
+/// `Error` that may be thrown if we fail to create a `IPAddress`
+public enum IPAddressError: Error {
+    /// Given string input is not supported IP Address
+    case failedToParseIPString(String)
+    /// Given string input is not supported IP Address
+    case bytesArrayHasWrongLength(Int)
+}
+
+public typealias IPv4BytesTuple = (UInt8, UInt8, UInt8, UInt8)
+public typealias IPv6BytesTuple = (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8)
+
+/// Represent the bytes for an `IPv4Address`.
+public struct IPv4Bytes {
+    private let ipv4BytesTuple: IPv4BytesTuple
```

Comment:
For both this and the `IPv6Bytes` struct, this should be `var` and should probably be `public`. We can even make it the `bytes` accessor and simply remove that. Additionally, the `init` should be `public` for both types.